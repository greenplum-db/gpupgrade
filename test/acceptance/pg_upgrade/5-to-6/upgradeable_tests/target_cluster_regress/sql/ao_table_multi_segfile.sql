-- Copyright (c) 2017-2023 VMware, Inc. or its affiliates
-- SPDX-License-Identifier: Apache-2.0

--------------------------------------------------------------------------------
-- Validate that the upgradeable objects are functional post-upgrade
--------------------------------------------------------------------------------

-- Verify AO table with multiple segment files
SELECT * FROM ao_multi_segment ORDER BY id;
SELECT * FROM gp_toolkit.__gp_aoseg('ao_multi_segment');
SELECT gp_segment_id, (t).* FROM (SELECT gp_segment_id, gp_toolkit.__gp_aoseg('ao_multi_segment') AS t FROM gp_dist_random('gp_id')) AS x;

1:BEGIN;
2:BEGIN;
1:INSERT INTO ao_multi_segment VALUES (4, 'Jude');
1:INSERT INTO ao_multi_segment VALUES (5, 'Jade');
2:INSERT INTO ao_multi_segment VALUES (6, 'Jack');
1:END;
2:END;

SELECT * FROM ao_multi_segment ORDER BY id;
SELECT * FROM gp_toolkit.__gp_aoseg('ao_multi_segment');
SELECT gp_segment_id, (t).* FROM (SELECT gp_segment_id, gp_toolkit.__gp_aoseg('ao_multi_segment') AS t FROM gp_dist_random('gp_id')) AS x;



-- Verify compaction success after upgrade
SELECT * FROM ao_vacuum_compact_after_upgrade ORDER by a;
SELECT * FROM gp_toolkit.__gp_aoseg('ao_vacuum_compact_after_upgrade');
SELECT gp_segment_id, (t).* FROM (SELECT gp_segment_id, gp_toolkit.__gp_aoseg('ao_vacuum_compact_after_upgrade') AS t FROM gp_dist_random('gp_id')) AS x;

SET gp_select_invisible = on;
SELECT * FROM ao_vacuum_compact_after_upgrade ORDER BY a;
VACUUM ao_vacuum_compact_after_upgrade;
SELECT * FROM ao_vacuum_compact_after_upgrade ORDER BY a;
SET gp_select_invisible = off;

SELECT * FROM ao_vacuum_compact_after_upgrade ORDER BY a;
SELECT * FROM gp_toolkit.__gp_aoseg('ao_vacuum_compact_after_upgrade');
SELECT gp_segment_id, (t).* FROM (SELECT gp_segment_id, gp_toolkit.__gp_aoseg('ao_vacuum_compact_after_upgrade') AS t FROM gp_dist_random('gp_id')) AS x;



-- Verify compaction success after upgrade with awaiting drop state
SELECT * FROM ao_with_awaiting_drop_state_before_upgrade ORDER BY a;
SELECT * FROM gp_toolkit.__gp_aoseg('ao_with_awaiting_drop_state_before_upgrade');
SELECT gp_segment_id, (t).* FROM (SELECT gp_segment_id, gp_toolkit.__gp_aoseg('ao_with_awaiting_drop_state_before_upgrade') AS t FROM gp_dist_random('gp_id')) AS x;

VACUUM ao_with_awaiting_drop_state_before_upgrade;

SELECT * FROM gp_toolkit.__gp_aoseg('ao_with_awaiting_drop_state_before_upgrade');
SELECT gp_segment_id, (t).* FROM (SELECT gp_segment_id, gp_toolkit.__gp_aoseg('ao_with_awaiting_drop_state_before_upgrade') AS t FROM gp_dist_random('gp_id')) AS x;

1: BEGIN;
2: BEGIN;
1: INSERT INTO ao_with_awaiting_drop_state_before_upgrade VALUES (88, 88);
2: INSERT INTO ao_with_awaiting_drop_state_before_upgrade VALUES (88, 88);
3: INSERT INTO ao_with_awaiting_drop_state_before_upgrade VALUES (88, 88);
1: COMMIT;
2: COMMIT;

SELECT * FROM ao_with_awaiting_drop_state_before_upgrade ORDER BY a;
SELECT * FROM gp_toolkit.__gp_aoseg('ao_with_awaiting_drop_state_before_upgrade');
SELECT gp_segment_id, (t).* FROM (SELECT gp_segment_id, gp_toolkit.__gp_aoseg('ao_with_awaiting_drop_state_before_upgrade') AS t FROM gp_dist_random('gp_id')) AS x;



-- Verify empty AO segments are still there and can be inserted into
SELECT * FROM gp_toolkit.__gp_aoseg('ao_with_empty_aosegs_before_upgrade');
SELECT gp_segment_id, (t).* FROM (SELECT gp_segment_id, gp_toolkit.__gp_aoseg('ao_with_empty_aosegs_before_upgrade') AS t FROM gp_dist_random('gp_id')) AS x;
SELECT * FROM ao_with_empty_aosegs_before_upgrade;

INSERT INTO ao_with_empty_aosegs_before_upgrade SELECT generate_series(1,10);

SELECT * FROM ao_with_empty_aosegs_before_upgrade ORDER BY a;
SELECT * FROM gp_toolkit.__gp_aoseg('ao_with_empty_aosegs_before_upgrade');
SELECT gp_segment_id, (t).* FROM (SELECT gp_segment_id, gp_toolkit.__gp_aoseg('ao_with_empty_aosegs_before_upgrade') AS t FROM gp_dist_random('gp_id')) AS x;
