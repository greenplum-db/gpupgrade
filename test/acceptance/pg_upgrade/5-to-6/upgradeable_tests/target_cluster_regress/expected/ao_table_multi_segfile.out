-- Copyright (c) 2017-2023 VMware, Inc. or its affiliates
-- SPDX-License-Identifier: Apache-2.0

--------------------------------------------------------------------------------
-- Validate that the upgradeable objects are functional post-upgrade
--------------------------------------------------------------------------------

-- Verify AO table with multiple segment files
SELECT * FROM ao_multi_segment ORDER BY id;
 id | name 
----+------
 1  | Jane 
 2  | John 
 3  | Joe  
(3 rows)
SELECT * FROM gp_toolkit.__gp_aoseg('ao_multi_segment');
 segno | eof | tupcount | varblockcount | eof_uncompressed | modcount | formatversion | state 
-------+-----+----------+---------------+------------------+----------+---------------+-------
 0     | 0   | 0        | 0             | 0                | 1        | 3             | 1     
 1     | 0   | 2        | 0             | 0                | 2        | 3             | 1     
 2     | 0   | 1        | 0             | 0                | 1        | 3             | 1     
(3 rows)
SELECT gp_segment_id, (t).* FROM (SELECT gp_segment_id, gp_toolkit.__gp_aoseg('ao_multi_segment') AS t FROM gp_dist_random('gp_id')) AS x;
 gp_segment_id | segno | eof | tupcount | varblockcount | eof_uncompressed | modcount | formatversion | state 
---------------+-------+-----+----------+---------------+------------------+----------+---------------+-------
 0             | 1     | 80  | 2        | 2             | 80               | 2        | 3             | 1     
 1             | 2     | 40  | 1        | 1             | 40               | 1        | 3             | 1     
(2 rows)

1:BEGIN;
BEGIN
2:BEGIN;
BEGIN
1:INSERT INTO ao_multi_segment VALUES (4, 'Jude');
INSERT 1
1:INSERT INTO ao_multi_segment VALUES (5, 'Jade');
INSERT 1
2:INSERT INTO ao_multi_segment VALUES (6, 'Jack');
INSERT 1
1:END;
END
2:END;
END

SELECT * FROM ao_multi_segment ORDER BY id;
 id | name 
----+------
 1  | Jane 
 2  | John 
 3  | Joe  
 4  | Jude 
 5  | Jade 
 6  | Jack 
(6 rows)
SELECT * FROM gp_toolkit.__gp_aoseg('ao_multi_segment');
 segno | eof | tupcount | varblockcount | eof_uncompressed | modcount | formatversion | state 
-------+-----+----------+---------------+------------------+----------+---------------+-------
 0     | 0   | 0        | 0             | 0                | 1        | 3             | 1     
 1     | 0   | 4        | 0             | 0                | 4        | 3             | 1     
 2     | 0   | 2        | 0             | 0                | 2        | 3             | 1     
(3 rows)
SELECT gp_segment_id, (t).* FROM (SELECT gp_segment_id, gp_toolkit.__gp_aoseg('ao_multi_segment') AS t FROM gp_dist_random('gp_id')) AS x;
 gp_segment_id | segno | eof | tupcount | varblockcount | eof_uncompressed | modcount | formatversion | state 
---------------+-------+-----+----------+---------------+------------------+----------+---------------+-------
 0             | 1     | 80  | 2        | 2             | 80               | 2        | 3             | 1     
 1             | 1     | 80  | 2        | 2             | 80               | 2        | 3             | 1     
 1             | 2     | 80  | 2        | 2             | 80               | 2        | 3             | 1     
(3 rows)



-- Verify compaction success after upgrade
SELECT * FROM ao_vacuum_compact_after_upgrade ORDER by a;
 a | b 
---+---
 1 | 1 
 2 | 2 
 3 | 3 
 4 | 4 
 5 | 5 
(5 rows)
SELECT * FROM gp_toolkit.__gp_aoseg('ao_vacuum_compact_after_upgrade');
 segno | eof | tupcount | varblockcount | eof_uncompressed | modcount | formatversion | state 
-------+-----+----------+---------------+------------------+----------+---------------+-------
 0     | 0   | 0        | 0             | 0                | 1        | 3             | 1     
 1     | 0   | 10       | 0             | 0                | 2        | 3             | 1     
 2     | 0   | 10       | 0             | 0                | 1        | 3             | 1     
(3 rows)
SELECT gp_segment_id, (t).* FROM (SELECT gp_segment_id, gp_toolkit.__gp_aoseg('ao_vacuum_compact_after_upgrade') AS t FROM gp_dist_random('gp_id')) AS x;
 gp_segment_id | segno | eof | tupcount | varblockcount | eof_uncompressed | modcount | formatversion | state 
---------------+-------+-----+----------+---------------+------------------+----------+---------------+-------
 0             | 1     | 72  | 2        | 1             | 72               | 1        | 3             | 1     
 0             | 2     | 128 | 5        | 1             | 128              | 1        | 3             | 1     
 1             | 1     | 128 | 5        | 1             | 128              | 1        | 3             | 1     
 1             | 2     | 88  | 3        | 1             | 88               | 1        | 3             | 1     
 2             | 1     | 88  | 3        | 1             | 88               | 1        | 3             | 1     
 2             | 2     | 72  | 2        | 1             | 72               | 1        | 3             | 1     
(6 rows)

SET gp_select_invisible = on;
SET
SELECT * FROM ao_vacuum_compact_after_upgrade ORDER BY a;
 a  | b  
----+----
 1  | 1  
 2  | 2  
 3  | 3  
 4  | 4  
 5  | 5  
 6  | 6  
 7  | 7  
 8  | 8  
 9  | 9  
 10 | 10 
 11 | 11 
 12 | 12 
 13 | 13 
 14 | 14 
 15 | 15 
 16 | 16 
 17 | 17 
 18 | 18 
 19 | 19 
 20 | 20 
(20 rows)
VACUUM ao_vacuum_compact_after_upgrade;
VACUUM
SELECT * FROM ao_vacuum_compact_after_upgrade ORDER BY a;
 a | b 
---+---
 1 | 1 
 2 | 2 
 3 | 3 
 4 | 4 
 5 | 5 
(5 rows)
SET gp_select_invisible = off;
SET

SELECT * FROM ao_vacuum_compact_after_upgrade ORDER BY a;
 a | b 
---+---
 1 | 1 
 2 | 2 
 3 | 3 
 4 | 4 
 5 | 5 
(5 rows)
SELECT * FROM gp_toolkit.__gp_aoseg('ao_vacuum_compact_after_upgrade');
 segno | eof | tupcount | varblockcount | eof_uncompressed | modcount | formatversion | state 
-------+-----+----------+---------------+------------------+----------+---------------+-------
 0     | 0   | 0        | 0             | 0                | 1        | 3             | 1     
 1     | 0   | 18       | 0             | 0                | 2        | 3             | 1     
 2     | 0   | 20       | 0             | 0                | 1        | 3             | 1     
 3     | 0   | 3        | 0             | 0                | 0        | 3             | 1     
(4 rows)
SELECT gp_segment_id, (t).* FROM (SELECT gp_segment_id, gp_toolkit.__gp_aoseg('ao_vacuum_compact_after_upgrade') AS t FROM gp_dist_random('gp_id')) AS x;
 gp_segment_id | segno | eof | tupcount | varblockcount | eof_uncompressed | modcount | formatversion | state 
---------------+-------+-----+----------+---------------+------------------+----------+---------------+-------
 0             | 1     | 72  | 2        | 1             | 72               | 1        | 3             | 1     
 0             | 2     | 0   | 0        | 0             | 0                | 1        | 3             | 1     
 0             | 3     | 0   | 0        | 0             | 0                | 1        | 3             | 1     
 0             | 4     | 0   | 0        | 0             | 0                | 1        | 3             | 1     
 1             | 1     | 0   | 0        | 0             | 0                | 1        | 3             | 1     
 1             | 2     | 0   | 0        | 0             | 0                | 1        | 3             | 1     
 1             | 3     | 88  | 3        | 1             | 88               | 1        | 3             | 1     
 1             | 4     | 0   | 0        | 0             | 0                | 1        | 3             | 1     
 2             | 1     | 0   | 0        | 0             | 0                | 1        | 3             | 1     
 2             | 2     | 0   | 0        | 0             | 0                | 1        | 3             | 1     
 2             | 3     | 0   | 0        | 0             | 0                | 1        | 3             | 1     
 2             | 4     | 0   | 0        | 0             | 0                | 1        | 3             | 1     
(12 rows)



-- Verify compaction success after upgrade with awaiting drop state
SELECT * FROM ao_with_awaiting_drop_state_before_upgrade ORDER BY a;
 a  | b  
----+----
 1  | 1  
 2  | 2  
 3  | 3  
 4  | 4  
 5  | 5  
 6  | 6  
 7  | 7  
 8  | 8  
 9  | 9  
 10 | 10 
(10 rows)
SELECT * FROM gp_toolkit.__gp_aoseg('ao_with_awaiting_drop_state_before_upgrade');
 segno | eof | tupcount | varblockcount | eof_uncompressed | modcount | formatversion | state 
-------+-----+----------+---------------+------------------+----------+---------------+-------
 0     | 0   | 0        | 0             | 0                | 1        | 3             | 1     
 1     | 0   | 10       | 0             | 0                | 2        | 3             | 1     
 2     | 0   | 10       | 0             | 0                | 1        | 3             | 1     
(3 rows)
SELECT gp_segment_id, (t).* FROM (SELECT gp_segment_id, gp_toolkit.__gp_aoseg('ao_with_awaiting_drop_state_before_upgrade') AS t FROM gp_dist_random('gp_id')) AS x;
 gp_segment_id | segno | eof | tupcount | varblockcount | eof_uncompressed | modcount | formatversion | state 
---------------+-------+-----+----------+---------------+------------------+----------+---------------+-------
 0             | 1     | 72  | 2        | 1             | 72               | 1        | 3             | 2     
 0             | 2     | 72  | 2        | 1             | 72               | 2        | 3             | 1     
 1             | 1     | 128 | 5        | 1             | 128              | 1        | 3             | 2     
 1             | 2     | 128 | 5        | 1             | 128              | 2        | 3             | 1     
 2             | 1     | 88  | 3        | 1             | 88               | 1        | 3             | 2     
 2             | 2     | 88  | 3        | 1             | 88               | 2        | 3             | 1     
(6 rows)

VACUUM ao_with_awaiting_drop_state_before_upgrade;
VACUUM

SELECT * FROM gp_toolkit.__gp_aoseg('ao_with_awaiting_drop_state_before_upgrade');
 segno | eof | tupcount | varblockcount | eof_uncompressed | modcount | formatversion | state 
-------+-----+----------+---------------+------------------+----------+---------------+-------
 0     | 0   | 0        | 0             | 0                | 1        | 3             | 1     
 1     | 0   | 0        | 0             | 0                | 2        | 3             | 1     
 2     | 0   | 10       | 0             | 0                | 1        | 3             | 1     
(3 rows)
SELECT gp_segment_id, (t).* FROM (SELECT gp_segment_id, gp_toolkit.__gp_aoseg('ao_with_awaiting_drop_state_before_upgrade') AS t FROM gp_dist_random('gp_id')) AS x;
 gp_segment_id | segno | eof | tupcount | varblockcount | eof_uncompressed | modcount | formatversion | state 
---------------+-------+-----+----------+---------------+------------------+----------+---------------+-------
 0             | 1     | 0   | 0        | 0             | 0                | 1        | 3             | 1     
 0             | 2     | 72  | 2        | 1             | 72               | 2        | 3             | 1     
 0             | 3     | 0   | 0        | 0             | 0                | 1        | 3             | 1     
 1             | 1     | 0   | 0        | 0             | 0                | 1        | 3             | 1     
 1             | 2     | 128 | 5        | 1             | 128              | 2        | 3             | 1     
 1             | 3     | 0   | 0        | 0             | 0                | 1        | 3             | 1     
 2             | 1     | 0   | 0        | 0             | 0                | 1        | 3             | 1     
 2             | 2     | 88  | 3        | 1             | 88               | 2        | 3             | 1     
 2             | 3     | 0   | 0        | 0             | 0                | 1        | 3             | 1     
(9 rows)

1: BEGIN;
BEGIN
2: BEGIN;
BEGIN
1: INSERT INTO ao_with_awaiting_drop_state_before_upgrade VALUES (88, 88);
INSERT 1
2: INSERT INTO ao_with_awaiting_drop_state_before_upgrade VALUES (88, 88);
INSERT 1
3: INSERT INTO ao_with_awaiting_drop_state_before_upgrade VALUES (88, 88);
INSERT 1
1: COMMIT;
COMMIT
2: COMMIT;
COMMIT

SELECT * FROM ao_with_awaiting_drop_state_before_upgrade ORDER BY a;
 a  | b  
----+----
 1  | 1  
 2  | 2  
 3  | 3  
 4  | 4  
 5  | 5  
 6  | 6  
 7  | 7  
 8  | 8  
 9  | 9  
 10 | 10 
 88 | 88 
 88 | 88 
 88 | 88 
(13 rows)
SELECT * FROM gp_toolkit.__gp_aoseg('ao_with_awaiting_drop_state_before_upgrade');
 segno | eof | tupcount | varblockcount | eof_uncompressed | modcount | formatversion | state 
-------+-----+----------+---------------+------------------+----------+---------------+-------
 0     | 0   | 0        | 0             | 0                | 1        | 3             | 1     
 1     | 0   | 1        | 0             | 0                | 3        | 3             | 1     
 2     | 0   | 11       | 0             | 0                | 2        | 3             | 1     
 3     | 0   | 1        | 0             | 0                | 1        | 3             | 1     
(4 rows)
SELECT gp_segment_id, (t).* FROM (SELECT gp_segment_id, gp_toolkit.__gp_aoseg('ao_with_awaiting_drop_state_before_upgrade') AS t FROM gp_dist_random('gp_id')) AS x;
 gp_segment_id | segno | eof | tupcount | varblockcount | eof_uncompressed | modcount | formatversion | state 
---------------+-------+-----+----------+---------------+------------------+----------+---------------+-------
 0             | 1     | 0   | 0        | 0             | 0                | 1        | 3             | 1     
 0             | 2     | 72  | 2        | 1             | 72               | 2        | 3             | 1     
 0             | 3     | 0   | 0        | 0             | 0                | 1        | 3             | 1     
 1             | 1     | 40  | 1        | 1             | 40               | 2        | 3             | 1     
 1             | 2     | 168 | 6        | 2             | 168              | 3        | 3             | 1     
 1             | 3     | 40  | 1        | 1             | 40               | 2        | 3             | 1     
 2             | 1     | 0   | 0        | 0             | 0                | 1        | 3             | 1     
 2             | 2     | 88  | 3        | 1             | 88               | 2        | 3             | 1     
 2             | 3     | 0   | 0        | 0             | 0                | 1        | 3             | 1     
(9 rows)



-- Verify empty AO segments are still there and can be inserted into
SELECT * FROM gp_toolkit.__gp_aoseg('ao_with_empty_aosegs_before_upgrade');
 segno | eof | tupcount | varblockcount | eof_uncompressed | modcount | formatversion | state 
-------+-----+----------+---------------+------------------+----------+---------------+-------
 0     | 0   | 0        | 0             | 0                | 1        | 3             | 1     
 1     | 0   | 0        | 0             | 0                | 2        | 3             | 1     
 2     | 0   | 0        | 0             | 0                | 1        | 3             | 1     
 3     | 0   | 0        | 0             | 0                | 1        | 3             | 1     
(4 rows)
SELECT gp_segment_id, (t).* FROM (SELECT gp_segment_id, gp_toolkit.__gp_aoseg('ao_with_empty_aosegs_before_upgrade') AS t FROM gp_dist_random('gp_id')) AS x;
 gp_segment_id | segno | eof | tupcount | varblockcount | eof_uncompressed | modcount | formatversion | state 
---------------+-------+-----+----------+---------------+------------------+----------+---------------+-------
 0             | 1     | 0   | 0        | 0             | 0                | 1        | 3             | 1     
 0             | 2     | 0   | 0        | 0             | 0                | 1        | 3             | 1     
 0             | 3     | 0   | 0        | 0             | 0                | 1        | 3             | 1     
 0             | 4     | 0   | 0        | 0             | 0                | 3        | 3             | 1     
 1             | 1     | 0   | 0        | 0             | 0                | 1        | 3             | 1     
 1             | 2     | 0   | 0        | 0             | 0                | 1        | 3             | 1     
 1             | 3     | 0   | 0        | 0             | 0                | 1        | 3             | 1     
 1             | 4     | 0   | 0        | 0             | 0                | 3        | 3             | 1     
 2             | 1     | 0   | 0        | 0             | 0                | 1        | 3             | 1     
 2             | 2     | 0   | 0        | 0             | 0                | 1        | 3             | 1     
 2             | 3     | 0   | 0        | 0             | 0                | 1        | 3             | 1     
 2             | 4     | 0   | 0        | 0             | 0                | 3        | 3             | 1     
(12 rows)
SELECT * FROM ao_with_empty_aosegs_before_upgrade;
 a 
---
(0 rows)

INSERT INTO ao_with_empty_aosegs_before_upgrade SELECT generate_series(1,10);
INSERT 10

SELECT * FROM ao_with_empty_aosegs_before_upgrade ORDER BY a;
 a  
----
 1  
 2  
 3  
 4  
 5  
 6  
 7  
 8  
 9  
 10 
(10 rows)
SELECT * FROM gp_toolkit.__gp_aoseg('ao_with_empty_aosegs_before_upgrade');
 segno | eof | tupcount | varblockcount | eof_uncompressed | modcount | formatversion | state 
-------+-----+----------+---------------+------------------+----------+---------------+-------
 0     | 0   | 0        | 0             | 0                | 1        | 3             | 1     
 1     | 0   | 10       | 0             | 0                | 3        | 3             | 1     
 2     | 0   | 0        | 0             | 0                | 1        | 3             | 1     
 3     | 0   | 0        | 0             | 0                | 1        | 3             | 1     
(4 rows)
SELECT gp_segment_id, (t).* FROM (SELECT gp_segment_id, gp_toolkit.__gp_aoseg('ao_with_empty_aosegs_before_upgrade') AS t FROM gp_dist_random('gp_id')) AS x;
 gp_segment_id | segno | eof | tupcount | varblockcount | eof_uncompressed | modcount | formatversion | state 
---------------+-------+-----+----------+---------------+------------------+----------+---------------+-------
 0             | 1     | 56  | 2        | 1             | 56               | 2        | 3             | 1     
 0             | 2     | 0   | 0        | 0             | 0                | 1        | 3             | 1     
 0             | 3     | 0   | 0        | 0             | 0                | 1        | 3             | 1     
 0             | 4     | 0   | 0        | 0             | 0                | 3        | 3             | 1     
 1             | 1     | 88  | 5        | 1             | 88               | 2        | 3             | 1     
 1             | 2     | 0   | 0        | 0             | 0                | 1        | 3             | 1     
 1             | 3     | 0   | 0        | 0             | 0                | 1        | 3             | 1     
 1             | 4     | 0   | 0        | 0             | 0                | 3        | 3             | 1     
 2             | 1     | 64  | 3        | 1             | 64               | 2        | 3             | 1     
 2             | 2     | 0   | 0        | 0             | 0                | 1        | 3             | 1     
 2             | 3     | 0   | 0        | 0             | 0                | 1        | 3             | 1     
 2             | 4     | 0   | 0        | 0             | 0                | 3        | 3             | 1     
(12 rows)
